### ISSUE TITLE
ospSetProgressFunc not suitable for concurrent rendering

### ISSUE THREAD

**Issue by tribal-tec** [ospray/ospray#278]
Hi,

I was excited when I saw that new feature in the latest release of OSPRay, but now while I'm trying to use it in [Brayns](https://github.com/BlueBrain/Brayns.git), I realized that the callback has no notion of which renderer it belongs to.

Our software does use concurrent rendering (like 1 'live' renderer, plus optional snapshot renderers which use temporary renderers + framebuffers until the snapshot is done).

With the current state of the progress func, we would have to serialize calls to `ospRenderFrame()`, to set/reset the progress callback, where today we don't have to as TBB and everything else just works with concurrent rendering.

One sidenote: We're mainly interested in the cancel feature of the progress callback :) It will allow for a better integration of cancelling snapshots, where today we call #spp times `ospRenderFrame()` in our own loop, which is less optimal. And it will make the interactive more responsive, if a high spp was set by the user.

Any chance you are thinking of making the progress func more context aware (renderer, framebuffer, ...) rather being on the device level? Looking at where it's used, if in the framebuffer, it should work for the local loadbalancer and distributed framebuffer case.

Looking forward to your opinion.
Cheers,
Daniel

**Comment by jeffamstutz** [comment]
Hey Daniel,

The core issue is that the OSPRay API is not entirely thread safe, nor are OSPRay render calls asynchronous. We have some of the exact same desires as you as we also do rendering concurrently with the GUI thread in our example viewer (which is moving to the new OSPRay Studio project, btw).

Thus the callback feature is correctly implemented given the current limitations on the API. However, we do intend to look at concurrent rendering after SC as a part of API changes for v2.0+. I actually think getting the API-side correct will not be terribly difficult: all we need is something like an ```OSPFuture``` handle returned from ```ospRenderFrame()```, which can be passed to functions to synchronize/cancel/query/etc. In that context, getting the callback specified correctly will be the easy part. :)

I'm not sure if we will make global synchronization guarantees between object ```commit()``` calls and ```ospRenderFrame()``` (that's an entirely different beast), but concurrent rendering needs to be explicitly handled to implement the desired behavior. Knowing an external user would greatly benefit from such a feature greatly increases its priority. Fortunately, I think we can add new API calls to implement it in a future v1.x release (that don't break existing API usage) so you don't have to wait until all of v2.0 is done.

Lastly, we will need to think about how this will work with our MPI backends...I don't anticipate any major issues, but MPI can be...well...moody sometimes. :)

**Comment by jeffamstutz** [comment]
UPDATE: a prototype implementation of the async API is underway internally and already showing promise. We will make sure this use case gets covered to allow cancelling specific frames.

**Comment by tribal-tec** [comment]
That's good, looking forward to try that it out soon.

**Comment by jeffamstutz** [comment]
The first pass of the async API was merged this morning, so it is now in `devel` for you to try out!

The following API calls have been added:

`ospRenderFrameAsync`
`ospIsReady`
`ospWait`
`ospCancel`
`ospGetProgress`
`ospGetVariance`

We have removed the callback method of getting progress (push approach) in favor of explicit functions to poll status/info. This is not final, but it does simplify the complexity required to inject which frame the callback is invoked on (multiple may be in flight).
